Use Ref for CSSValue in CSSProperty

weinig opened the pull request:
#### a88754dd0e28e5c3df0d4004a452ece1eb59ee2f
<pre>
Use Ref for CSSValue in CSSProperty
<a href="https://bugs.webkit.org/show_bug.cgi?id=287105">https://bugs.webkit.org/show_bug.cgi?id=287105</a>

Reviewed by Tim Nguyen and Ryosuke Niwa.

The CSSValue in CSSProperty should never be null. Use a Ref
as the member and update all callers to pass in a Ref.

* Source/WebCore/css/CSSFontFace.cpp:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/MutableStyleProperties.cpp:
* Source/WebCore/css/MutableStyleProperties.h:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/parser/CSSParser.cpp:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/dom/StyledElement.cpp:
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/dom/ViewTransition.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/FontAttributeChanges.cpp:
* Source/WebCore/html/HTMLFontElement.cpp:
* Source/WebCore/html/HTMLHRElement.cpp:
* Source/WebCore/svg/SVGFontFaceElement.cpp:

Canonical link: <a href="https://commits.webkit.org/290105@main">https://commits.webkit.org/290105@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7def79edc142f595597c0ebb5277be3add46d240

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8528 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43625 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39764 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91055 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8915 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16712 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/93976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/26250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92006 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6806 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/80594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6555 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38872 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76917 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95815 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16184 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11827 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16440 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/76739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18907 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/21134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16198 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21509 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15939 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19390 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -655,10 +655,14 @@ void ViewTransition::setupDynamicStyleSheet(const AtomString& name, const Captur
+    if (auto transform = capturedElement.oldProperties->getPropertyCSSValue(CSSPropertyTransform))
+        props->setProperty(CSSPropertyTransform, transform.releaseNonNull());
+    if (auto filter = capturedElement.oldProperties->getPropertyCSSValue(CSSPropertyBackdropFilter))
+        props->setProperty(CSSPropertyBackdropFilter, filter.releaseNonNull());
```
Maybe this should use:

```c++
Ref props = capturedElement.oldProperties->copyProperties({CSSPropertyWidth,  CSSPropertyHeight, CSSPropertyTransform, CSSPropertyBackdropFilter});
```

instead?

nt1m replied:
Seems roughly equivalent and cleaner, I would switch to that.

rniwa reviewed (approved):
r=me either with copyProperties change or without.

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -654,13 +654,13 @@ void ViewTransition::setupDynamicStyleSheet(const AtomString& name, const Captur
-    props->setProperty(CSSPropertyBackdropFilter, capturedElement.oldProperties->getPropertyCSSValue(CSSPropertyBackdropFilter));
-
-    Ref keyframe = StyleRuleKeyframe::create(WTFMove(props));
+    CSSPropertyID keyframeProperties[] = {
```
Can we use `auto` here?

darinadler commented on Source/WebCore/css/StyleProperties.h:
```diff
@@ -60,9 +60,10 @@ class StyleProperties : public RefCounted<StyleProperties> {
 
         // FIXME: Remove this.
-        CSSProperty toCSSProperty() const { return CSSProperty(id(), const_cast<CSSValue*>(m_value), isImportant() ? IsImportant::Yes : IsImportant::No, m_metadata.m_isSetFromShorthand, m_metadata.m_indexInShorthandsVector, isImplicit()); }
+        CSSProperty toCSSProperty() const { return CSSProperty(id(), protectedValue(), isImportant() ? IsImportant::Yes : IsImportant::No, m_metadata.m_isSetFromShorthand, m_metadata.m_indexInShorthandsVector, isImplicit()); }
```
Why `protectedValue` here instead of `*value()`? I heard @cdumez say we try not to use those functions like this, but maybe we'd have to write `Ref` here otherwise and also I might be misremembering his remark/guideline?

Or maybe this is about our mixed up use of `const` here?

Seems like we should change `value()` to return a reference in a follow up.

darinadler commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -1354,7 +1354,7 @@ void EditingStyle::mergeStyle(const StyleProperties* style, CSSPropertyOverrideM
 
         if (mode == OverrideValues || (mode == DoNotOverrideValues && !value))
-            m_mutableStyle->setProperty(property.id(), property.value(), property.isImportant() ? IsImportant::Yes : IsImportant::No);
+            m_mutableStyle->setProperty(property.id(), property.protectedValue(), property.isImportant() ? IsImportant::Yes : IsImportant::No);
```
Seems like this should be `*value()`, not `protectedValue()`.

weinig committed:
Use Ref for CSSValue in CSSProperty
https://bugs.webkit.org/show_bug.cgi?id=287105

Reviewed by Tim Nguyen and Ryosuke Niwa.

The CSSValue in CSSProperty should never be null. Use a Ref
as the member and update all callers to pass in a Ref.

* Source/WebCore/css/CSSFontFace.cpp:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/MutableStyleProperties.cpp:
* Source/WebCore/css/MutableStyleProperties.h:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/parser/CSSParser.cpp:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/dom/StyledElement.cpp:
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/dom/ViewTransition.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/FontAttributeChanges.cpp:
* Source/WebCore/html/HTMLFontElement.cpp:
* Source/WebCore/html/HTMLHRElement.cpp:
* Source/WebCore/svg/SVGFontFaceElement.cpp:

Canonical link: https://commits.webkit.org/290105@main
